pkg/issuer/acme/dns/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,190 @@
+func NewDNSProvider() (*DNSProvider, error) {
+	token := os.Getenv("DIGITALOCEAN_TOKEN")
+	return NewDNSProviderCredentials(token)
+}
+
+// NewDNSProviderCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for digitalocean.
+func NewDNSProviderCredentials(token string) (*DNSProvider, error) {
+	if token == "" {
+		return nil, fmt.Errorf("DigitalOcean token missing")
+	}
+
+	return &DNSProvider{
+		authToken: token,
+	}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS
+// propagation. Adjusting here to cope with spikes in propagation times.
+func (c *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return 120 * time.Second, 2 * time.Second
+}
+
+// Present creates a TXT record to fulfil the dns-01 challenge
+func (c *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, _ := util.DNS01Record(domain, keyAuth)
+
+	// if DigitalOcean does not have this zone then we will find out later
+	zoneName, err := util.FindZoneByFqdn(fqdn, util.RecursiveNameservers)
+	if err != nil {
+		return err
+	}
+
+	rec := digitalOceanRecord{
+		Type: "TXT",
+		Name: util.UnFqdn(fqdn),
+		Data: value,
+		TTL:  120,
+	}
+
+	body, err := json.Marshal(rec)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.makeRequest("POST", fmt.Sprintf("/domains/%s/records", util.UnFqdn(zoneName)), bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CleanUp removes the TXT record matching the specified parameters
+func (c *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, _, _ := util.DNS01Record(domain, keyAuth)
+
+	zoneName, err := util.FindZoneByFqdn(fqdn, util.RecursiveNameservers)
+
+	record, err := c.findTxtRecord(fqdn)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.makeRequest("DELETE", fmt.Sprintf("/domains/%s/records/%d", util.UnFqdn(zoneName), record.ID), nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *DNSProvider) findTxtRecord(fqdn string) (*digitalOceanRecord, error) {
+
+	// APIDomainRecordsResponse represents a domain records list response from the digitalocean API
+	type APIDomainRecordsResponse struct {
+		DomainRecords []digitalOceanRecord `json:"domain_records"`
+	}
+
+	zoneName, err := util.FindZoneByFqdn(fqdn, util.RecursiveNameservers)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := c.makeRequest(
+		"GET",
+		fmt.Sprintf("/domains/%s/records?per_page=1000&type=TXT&name=%s", util.UnFqdn(zoneName), util.UnFqdn(fqdn)),
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var records APIDomainRecordsResponse
+	err = json.Unmarshal(result, &records)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rec := range records.DomainRecords {
+		if rec.Name == util.UnFqdn(fqdn) {
+			return &rec, nil
+		}
+	}
+
+	return nil, fmt.Errorf("No existing record found for %s", fqdn)
+}
+
+func (c *DNSProvider) makeRequest(method, uri string, body io.Reader) (json.RawMessage, error) {
```

Comment:
Any reason for hand-writing the 50 lines of client code we need vs using the digital ocean [godo](https://github.com/digitalocean/godo) library to make these calls?

Reply:
I am aware of the godo library. However, it is a library for the entire DO API, vs just 50 lines to make the 1 or 2 API calls needed. I was also observing the convention of the other DNS providers making similar choices.

Reply:
I think the other DNS providers all use the cloud-provider provided library, excepting cloudflare.

I'm personally in favour of using those libraries when possible since auth and retry logic can be a bit tricky otherwise.
I'll defer to @munnerz if he wishes to weigh in differently.

Reply:
+1 using the official go library reduces the amount of testing and maintenance for this area of code

Reply:
I'm open to doing this in future - so far, we've copied across all the DNS providers from xenolf/lego when implementing them here. There's nothing wrong with us changing/improving them, but I don't think it's a requirement in order to accept it initially.